Add concepts section to "split markets miners" doc (#996)
filecoin-project/filecoin-docs: docs/mine/lotus/split-markets-miners.md

{{ $frontmatter.description }}

[[TOC]]

## Concepts

Lotus performs mining operations, such as sealing files into sectors, calculating proofs over
those files and submitting the proofs on chain.
Lotus also performs markets operations, providing storage and serving retrievals to clients.

> It is now possible to run mining and markets operations in separate processes. Service providers can

Reviewer: ```suggestion
It is now possible to run mining and markets subsystems in separate processes. Service providers can
```

Reviewer: ```suggestion
It is possible to run mining and markets operations in separate processes. Service providers can
```

Woudl remove the temporality, since it'll be become outdated by the next release.

accept storage and retrieval deals without impacting ongoing mining operations.

> It is recommended to run the mining and markets processes on separate machines so that

Reviewer: ```suggestion
It is recommended to run the mining and markets processes on separate physical or virtual machines so that
```

Reviewer: We need to stress the difference between *process* and *machines*. When I first read it, I was a bit confused.

{{ $frontmatter.description }}

[[TOC]]

## Concepts

Lotus performs mining operations, such as sealing files into sectors, calculating proofs over
those files and submitting the proofs on chain.
Lotus also performs markets operations, providing storage and serving retrievals to clients.

It is now possible to run mining and markets subsystems in separate processes. Service providers can
accept storage and retrieval deals without impacting ongoing mining operations. The markets process

> communicates with the mining process over JSON-RPC.

Reviewer 2: Also mention how it communicates with the network?

#  EnableMarkets = true
```

These options are set to `true` by default. Until now, the monolith `lotus-miner` process
has been responsible for all functionality.

> These `subsystems` are designed to be grouped into two distinct types of `lotus-miner` nodes:

Reviewer: ```suggestion
Although `subsystems` can be enabled and disabled individually per process, there are only two combinations that are supported right now:
```

  EnableMarkets = false
```

When a `lotus-miner` instance is configured as a `markets` node, it exposes
a libp2p interface. The libp2p ports should be publically available so that
the node can be dialed by clients that wish to make storage and retrieval deals.

> 

Reviewer: ```suggestion

This guide won't go into details on how to make the endpoint connectable, as this ultimately depends on your network infrastructure and equipment. Generally speaking, configuring static port mappings on your external-facing router, and ensuring your markets node has an internal static IP address and port should be sufficient, in addition to adjusting any firewalls that may be present.

```

When a `lotus-miner` instance is configured as a `mining/sealing/proving` node,
it does not receive requests and should not be publicly exposed on the Internet.

> The `markets` node communicates with the `mining/sealing/proving` node via its
> JSON RPC interface.

Reviewer: ```suggestion
**The `markets` node communicates with the `mining/sealing/proving` node via its
JSON RPC interface.** Just to reiterate, the `mining/sealing/proving` no longer runs a libp2p interface, and the `markets` and `mining/sealing/proving` do not communicate over libp2p, but through HTTP and/or Websockets (JSON-RPC interface).
```

### Split the processes

#### Pre-requisites

Before splitting the markets service process from the monolith miner process,

> you should backup your miner's metadata repository. Start the `lotus-miner`

Reviewer: ```suggestion
you should backup your miner's metadata repository. Stop your currently running `lotus-miner`
```

and start them again with...

Reviewer: ^^ Let's assume that the miner is operational.

```shell
export LOTUS_BACKUP_BASE_PATH=/tmp
lotus daemon
```

#### 1. Restart the `mining/sealing/proving` miner process without the markets subsystem

1. Update your `config.toml` to set the `EnableMarkets` option to `false`.

2. Stop the lotus miner, then restart it with the `LOTUS_BACKUP_BASE_PATH` environment variable:

```shell
export LOTUS_BACKUP_BASE_PATH=/tmp
lotus-miner run
```

The miner is now running as a `mining/sealing/proving` node (without the `markets` subsystem) 

#### 2. Create a backup

```shell
export LOTUS_BACKUP_BASE_PATH=/tmp
lotus-miner backup /tmp/backup.cbor
```

#### 3. Create a seed `config.toml` for the markets service

To initialize the markets service we need to create a seed `config.toml` for
the markets node. For more information see the [configuration page](https://docs.filecoin.io/get-started/lotus/configuration-and-advanced-usage/) and the [custom storage layout page](https://docs.filecoin.io/mine/lotus/custom-storage-layout/).

In the examples below the `config.toml` is in the `/tmp` directory. You may customise the example below to adapt it to your environment.
The repository directory is `~/.lotusmarkets`.

> We are using the following `config.toml`:

Reviewer 2: Still missing the [DealMaking] session

:: info
The `[Libp2p]` section of the `config.toml` on the `mining/sealing/proving` node
can be removed because it will no longer be running a Libp2p node, as we explained above.
::

### 3. Initialising a `markets` service repository
#### 4. Initialize a `markets` service repository

1. Create authentication tokens for the `markets` node. This is an online operation, so your `lotus-miner` should be running with its updated configuration (in case you move its API to another IP:PORT)
1. Create authentication tokens that will be used by the `markets` node to make JSON-RPC
calls to the `mining/sealing/proving` node.

```shell
export APISEALER=`lotus-miner auth api-info --perm=admin`
export APISECTORINDEX=`lotus-miner auth api-info --perm=admin`
```

2. Initialise the `market` node. This performs a one-time setup of the markets node. Part of that setup includes updating the `peer id` in the miner actor by submitting a message on chain. This is necessary so that storage and retrieval clients know that this miner's **deal-making** endpoint is now publicly dialable/reachable on a new address (the new `market` node).

> Note that `lotus-miner` interacts with one repository or another depending on the `LOTUS_MINER_PATH` environment variable!

Reviewer 3: How does this make sense if I'm going to initialize the markets node on a separate machine? In this case, 
LOTUS_MINER_PATH would be somewhere on another machine.

lotus-miner run
```

#### 1. Create a backup

```shell
export LOTUS_BACKUP_BASE_PATH=/tmp
lotus-miner backup /tmp/backup.cbor

> ```

Reviewer: Shouldn't this be part of the prerequisites?

Author: I don't think the word prerequisites really makes sense, as backing up the miner is an integral part of the process. I'm just going to remove the prerequisites heading

> 

Reviewer: Do we need to restart the nodes without `LOTUS_BACKUP_BASE_PATH`?

Author: I don't believe it matters

#### 2. Create a seed `config.toml` for the markets service

To initialize the markets service we need to create a seed `config.toml` for
the markets node. For more information see the [configuration page](https://docs.filecoin.io/get-started/lotus/configuration-and-advanced-usage/) and the [custom storage layout page](https://docs.filecoin.io/mine/lotus/custom-storage-layout/).

> In the examples below the `config.toml` is in the `/tmp` directory.

Reviewer: ```suggestion
In the examples below the `config.toml` is in the `/tmp` directory. You may customise the example below to adapt it to your environment.
```

The repository directory is `~/.lotusmarkets`.
We are using the following `config.toml`:

```toml
[API]

>   # Endpoint for API calls to the markets process

Reviewer: ```suggestion
  # Endpoint where the markets process will serve JSON-RPC requests
```

  ListenAddress = "/ip4/127.0.0.1/tcp/8787/http"
  RemoteListenAddress = "127.0.0.1:8787"

[Libp2p]
  # Endpoint for libp2p requests (public)
  ListenAddresses = ["/ip4/0.0.0.0/tcp/24001", "/ip6/::/tcp/24001"]
  AnnounceAddresses = ["/ip4/12.34.56.78/tcp/24001"]
```

> If you intend to run the `mining/sealing/proving` node on the same machine as the
> `markets` node, make sure that their listeners do not clash.

Reviewer: ```suggestion
If you intend to run the `mining/sealing/proving` node on the same machine as the
`markets` node, make sure that their `[API]` listener addresses do not clash.
```

Reviewer: Wrap it in a `::warning::`?

By default the `lotus-miner` API server listens to port 2345, so in the example

> configuration above, we change the configuration for the `markets` node API server

Reviewer: ```suggestion
configuration above for the `markets` process, we change the configuration for the `markets` node API server
```

to listen to `127.0.0.1:8787`.

Make sure you adjust the `[Libp2p]` section on the `markets` node accordingly - it
needs to be publicly accessible so that clients can make storage and retrieval
deals with your system.

> The `[Libp2p]` section of the `config.toml` on the `mining/sealing/proving` node
> can be removed because it will no longer be running a Libp2p node.

Reviewer: ```suggestion
:: info
The `[Libp2p]` section of the `config.toml` on the `mining/sealing/proving` node
can be removed because it will no longer be running a Libp2p node, as we explained above.
::
```

#### 3. Initialising a `markets` service repository

> 1. Create authentication tokens for the `markets` node. This is an online operation,

Reviewer: If that's the case, there should be an instruction to start the lotus markets process?

so your `lotus-miner` should be running with its updated configuration (in case

> you move its API to a different IP:PORT)

Reviewer: This bit is confusing :-/

Author: Yeah, it's true, need to think how to rework this part 🤔

```shell
export APISEALER=`lotus-miner auth api-info --perm=admin`
export APISECTORINDEX=`lotus-miner auth api-info --perm=admin`

> ```

Reviewer: There is no explanation for what this is?

Author: These params are passed to the init function below, not sure what they do exactly but I'll add a note that they're needed for init

> 2. Initialise the `market` node. This performs a one-time setup of the markets node.

Reviewer: Is it `market` or `markets`?

Part of that setup includes updating the `peer id` in the miner actor by submitting
a message on chain. This is necessary so that storage and retrieval clients know that
this miner's **deal-making** endpoint is now publicly dialable/reachable on a new
address (the new `market` node).

> This command should be run on the `markets` miner instance host, as it is creating

Reviewer: I think this is obvious, and this wording adds extra cognitive load because we haven't used the term "miner instance host" still...

Author: Removed

the `markets` miner instance repository, among other actions.

```shell

> lotus-miner --markets-repo=~/.lotusmarkets init service --type=markets \

Reviewer: We should probably change this command to `lotus-miner init subsystem`, if we're settling on subsystems as the canonical name for these. The term `service` hasn't appeared anywhere up until now...

                                                        /tmp/backup.cbor
```

> 3. Optionally update your miner's `multiaddr` on-chain - in case your `markets`

Reviewer: ```suggestion
3. If during this procedure, you ended up changing the `multiaddr` of your `markets` process to a different host and/or port, you must update miner's `multiaddr` on-chain. Otherwise, clients wishing to make deals with you won't find you any longer.
```

instance is publicly exposed at a different location compared to your existing
monolith miner, you also need to update your `multiaddr` on-chain to advertise
the correct address to clients:

```shell
lotus-miner actor set-addrs <NEW_MULTIADDR>
```

#### 4. Move the DAG store directory to the markets node repository

If you are running a lotus version with a DAG store, you can optionally move the
DAG store directory to the lotus markets repository, to avoid having to reindex
all storage deals when the markets node starts up.

```shell
mv ~/.lotusminer/dagStore ~/.lotusmarkets/

> ```

Reviewer: This needs to be optional, if the user had started 1.11.2-rc or m1.x with a monolith deployment.

Author: Are you saying we should add that phrasing here?

#### 5. Start the `mining/sealing/proving` miner process without the markets subsystem

1. Update your `config.toml` to set the `EnableMarkets` option to `false`.

2. Start the `mining/sealing/proving` node (with the default LOTUS_MINER_PATH, which

> should point to your `mining/sealing/proving` node repo).

Reviewer: During all this time the miner process was down? If that's the case, that's potentially significant downtime. We should either put a big warning at the top, or rejig the instructions so that they don't require so much downtime (e.g. we can change configurations in-place and simply restart the miner?)

Author: I think in practice you can just bounce your miner at this point - I believe the only time it needs to be down is when you do the backup. Would need to verify that with Anton though.

Author: Hmm actually the miner needs to be *up* when you do the backup. But at the same time you don't want to be accepting any deals as that would change state. So maybe the user should change their miner config first (so they disable the markets process), then bring the miner up, do the backup, and then bring up the markets node.

Reviewer: Makes sense. I would expect us to stop accepting deals, then backup, then stand up the new markets node and start accepting deals again. Otherwise the migration is not transactional and we risk the markets state slipping 👍 

One question: do we have commands to delete the markets state from the `mining/sealing/proving` mode, as it's now rendered irrelevant in its repo?

Author: > do we have commands to delete the markets state

I don't think so - and we also copy all that state over to the markets node

    Note that `lotus-miner` interacts with either the the `markets` or `mining/sealing/proving` repository
depending on the `LOTUS_MINER_PATH` environment variable!

```shell
lotus-miner run
```

#### 6. Start the `markets` miner process with the markets subsystem

> Start the `markets` node with LOTUS_MINER_PATH pointing to the `markets` node repo.

Reviewer: ```suggestion
Start the `markets` node with the environment variable `LOTUS_MINER_PATH` pointing to the `markets` node repo.
```

Reviewer: We need to explain the different environment variables involved at the top.

Note that `lotus-miner` interacts with either the the `markets` or `mining/sealing/proving` repository
depending on the `LOTUS_MINER_PATH` environment variable!

```shell

> LOTUS_MINER_PATH=~/.lotusmarkets lotus-miner run

Reviewer: We need to review the section below too.

                                                        /tmp/backup.cbor
```

3. If during this procedure, you ended up changing the `multiaddr` of your `markets`
process to a different host and/or port, you must update miner's `multiaddr` on-chain.
Otherwise, clients wishing to make deals with you will no longer be able to connect to
your node.

```shell
lotus-miner actor set-addrs <NEW_MULTIADDR>
```

#### 5. Move the DAG store directory to the markets node repository (optional)

If you are running a lotus version with a DAG store, you can optionally move the
DAG store directory to the lotus markets repository, to avoid having to reindex

> all storage deals when the markets node starts up.

Reviewer 3: Might include a comment on transferring the content of the directory, if markets node is going to be running on a separate machine.

```shell

> mv ~/.lotusminer/dagStore ~/.lotusmarkets/

Reviewer 3: After m1.3.4, this line should be `mv ~/.lotusminer/dagstore ~/.lotusmarkets/`
